registration/migrations/0007_fix_invalid_domain_names.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.7 on 2018-01-05 17:04
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.db.models import Func, F, Value
+from django.db.models.functions import Lower
+
+
+def clean_domain_name(field_name):
+    """
+    Returns a Django query expression that replaces all underscores with a
+    hyphen, removes leading and trailing hyphens, and converts the field to
+    lower case.
+    """
+    remove_underscores = Func(
+        F(field_name),
+        Value('_'),
+        Value('-'),
+        function='replace'
+    )
+    trim_hyphens = Func(
+        remove_underscores,
+        Value('-'),
+        function='btrim',
+    )
+    remove_trailing_hyphens = Func(
+        trim_hyphens,
+        Value(r'[-]+\.'),
+        Value('.'),
```

Comment:
No, actually. I've only ever seen it when managing DNS, but I'm pretty sure applications are always supposed to honor it, according to the RFCs. It makes the implicit root domain explicit. https://en.wikipedia.org/wiki/Fully_qualified_domain_name

Reply:
No I think it searches for "-." or "----." etc. and replaces it with ".", preserving the existing dot?  e.g. bad-.example.com becomes bad.example.com, not badexample.com

Reply:
Oh ah, gotcha.  Thanks for the clarification!

Reply:
Yes. I added this because the old regex would allow trailing hyphens or underscores, and the previous step would convert a trailing underscore to a hyphen. 

There is still a chance for collision where two users have subdomains like ``_bad_subdomain`` and ``bad_subdomain`` which would both resolve to ``bad-subdomain``, but I the main Ocim doesn't have any of those. 

Should I add code to handle those cases if we / others have additional Ocim deployments?

Reply:
@xitij2000 AFAIK, no one else runs Ocim yet.  We have a client who is considering it for their production deployments, but their domains would likely be very clean, so I wouldn't worry about handling collisions here.